File: Visual Basic.Net/Ginkgo GPIO/VB.NET_USB_GPIO_Test/USB_GPIO_Test/USB_GPIO_Test/ControlGPIO.vb
Option Strict Off
Option Explicit On

Imports System.Runtime.InteropServices
Module ControlGPIO
    'General Error Code
    Public Const SUCCESS As Int32 = 0                   ' no Error
    Public Const PARAMETER_NULL As Int32 = -1           ' void pointer 
    Public Const INPUT_DATA_TOO_MUCH As Int32 = -2      ' too many parameters
    Public Const INPUT_DATA_TOO_LESS As Int32 = -3      ' too few parameters
    Public Const INPUT_DATA_ILLEGALITY As Int32 = -4    ' illegal parameter
    Public Const USB_WRITE_DATA_ERROR As Int32 = -5     ' USB write data error
    Public Const USB_READ_DATA_ERROR As Int32 = -6      ' USB read data error
    Public Const READ_NO_DATA As Int32 = -7             ' no data return when request data
    Public Const OPEN_DEVICE_FAILD As Int32 = -8        ' failed to open device
    Public Const CLOSE_DEVICE_FAILD As Int32 = -9       ' failed to close device
    Public Const EXECUTE_CMD_FAILD As Int32 = -10       ' the command failed to execute  
    Public Const SELECT_DEVICE_FAILD As Int32 = -11     ' failed to select device
    Public Const DEVICE_OPENED As Int32 = -12           ' device has open
    Public Const DEVICE_NOTOPEN As Int32 = -13          ' device not open
    Public Const BUFFER_OVERFLOW As Int32 = -14         ' buffer overflow
    Public Const DEVICE_NOTEXIST As Int32 = -15         ' device not exist
    Public Const LOAD_KERNELDLL As Int32 = -16          ' failed to load KernelDLL
    Public Const CMD_FAILED As Int32 = -17              ' failed to execute command
    Public Const BUFFER_CREATE As Int32 = -18           ' out of memory
    ' Define Adapter type
    Public Const VGI_USBGPIO As Int32 = 1
    ' GPIO pin
    Public Const VGI_GPIO_PIN0 As UInt16 = 1 << 0       'GPIO_0
    Public Const VGI_GPIO_PIN1 As UInt16 = 1 << 1       'GPIO_1
    Public Const VGI_GPIO_PIN2 As UInt16 = 1 << 2       'GPIO_2
    Public Const VGI_GPIO_PIN3 As UInt16 = 1 << 3       'GPIO_3
    Public Const VGI_GPIO_PIN4 As UInt16 = 1 << 4       'GPIO_4
    Public Const VGI_GPIO_PIN5 As UInt16 = 1 << 5       'GPIO_5
    Public Const VGI_GPIO_PIN6 As UInt16 = 1 << 6       'GPIO_6
    Public Const VGI_GPIO_PIN7 As UInt16 = 1 << 7       'GPIO_7
    Public Const VGI_GPIO_PIN8 As UInt16 = 1 << 8       'GPIO_8
    Public Const VGI_GPIO_PIN9 As UInt16 = 1 << 9       'GPIO_9
    Public Const VGI_GPIO_PIN10 As UInt16 = 1 << 10     'GPIO_10
    Public Const VGI_GPIO_PIN11 As UInt16 = 1 << 11     'GPIO_11
    Public Const VGI_GPIO_PIN12 As UInt16 = 1 << 12     'GPIO_12
    Public Const VGI_GPIO_PIN13 As UInt16 = 1 << 13     'GPIO_13
    Public Const VGI_GPIO_PIN14 As UInt16 = 1 << 14     'GPIO_14
    Public Const VGI_GPIO_PIN15 As UInt16 = 1 << 15     'GPIO_15
    Public Const VGI_GPIO_PIN_ALL As UInt16 = &HFFFF    'GPIO_ALL
    ' Function
    Declare Function VGI_ScanDevice Lib "Ginkgo_Driver.dll" (ByVal NeedInit As Byte) As Int32
    Declare Function VGI_OpenDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal Reserved As Int32) As Int32
    Declare Function VGI_CloseDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32) As Int32
    ' Get pin value
    Declare Function VGI_ReadDatas Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16, ByRef pData As UInt16) As Int32
    ' Control pin output
    Declare Function VGI_WriteDatas Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16, ByVal Data As UInt16) As Int32
    Declare Function VGI_SetPins Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16) As Int32
    Declare Function VGI_ResetPins Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16) As Int32
    ' Set input/output and work mode
    Declare Function VGI_InitGPIO Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16, ByVal PinMode As UInt32) As Int32
    Declare Function VGI_SetInput Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16) As Int32
    Declare Function VGI_SetOutput Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16) As Int32
    Declare Function VGI_SetOpenDrain Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal PinMask As UInt16) As Int32

End Module

File: Delphi/Ginkgo GPIO/Delphi_USB_GPIO_Test/ControlGPIO.pas
unit ControlGPIO;

interface
const
//Device type definition
VAI_USBADC	=	1;
VCI_USBCAN1	=	3;
VCI_USBCAN2	=	4;
VGI_USBGPIO	=	1;
VII_USBI2C	=	1;
VSI_USBSPI	=	2;
//错误码定义
ERR_SUCCESS					      =	(0);		// 没有错误
ERR_PARAMETER_NULL			  =	(-1);	  // 传入的指针为空指针
ERR_INPUT_DATA_TOO_MUCH		=	(-2);	  // 参数输入个数多余规定个数
ERR_INPUT_DATA_TOO_LESS		=	(-3);	  // 参数输入个数少余规定个数
ERR_INPUT_DATA_ILLEGALITY	=	(-4);	  // 参数传入格式和规定的不符合
ERR_USB_WRITE_DATA			  =	(-5);	  // USB写数据错误
ERR_USB_READ_DATA			    =	(-6);	  // USB读数据错误
ERR_READ_NO_DATA			    =	(-7);	  // 请求读数据时返回没有数据
ERR_OPEN_DEVICE				    =	(-8);	  // 打开设备失败
ERR_CLOSE_DEVICE			    =	(-9);	  // 关闭设备失败
ERR_EXECUTE_CMD				    =	(-10);	// 设备执行命令失败
ERR_SELECT_DEVICE			    =	(-11);	// 选择设备失败
ERR_DEVICE_OPENED			    =	(-12);	// 设备已经打开
ERR_DEVICE_NOTOPEN			  =	(-13);	// 设备没有打开
ERR_BUFFER_OVERFLOW			  =	(-14);	// 缓冲区溢出
ERR_DEVICE_NOTEXIST			  =	(-15);	// 此设备不存在
ERR_LOAD_KERNELDLL			  =	(-16);	// 装载动态库失败
ERR_CMD_FAILED				    =	(-17);	// 执行命令失败错误码
ERR_BUFFER_CREATE			    =	(-18);	// 内存不足
//定义GPIO引脚号
VGI_GPIO_PIN0		= (1 shl 0);	//GPIO_0
VGI_GPIO_PIN1		= (1 shl 1);	//GPIO_1
VGI_GPIO_PIN2		= (1 shl 2);	//GPIO_2
VGI_GPIO_PIN3		= (1 shl 3);	//GPIO_3
VGI_GPIO_PIN4		= (1 shl 4);	//GPIO_4
VGI_GPIO_PIN5		= (1 shl 5);	//GPIO_5
VGI_GPIO_PIN6		= (1 shl 6);	//GPIO_6
VGI_GPIO_PIN7		= (1 shl 7);	//GPIO_7
VGI_GPIO_PIN8		= (1 shl 8);	//GPIO_1
VGI_GPIO_PIN9		= (1 shl 9);	//GPIO_2
VGI_GPIO_PIN10	= (1 shl 10);	//GPIO_3
VGI_GPIO_PIN11	= (1 shl 11);	//GPIO_4
VGI_GPIO_PIN12	= (1 shl 12);	//GPIO_5
VGI_GPIO_PIN13	= (1 shl 13);	//GPIO_6
VGI_GPIO_PIN14	= (1 shl 14);	//GPIO_7
VGI_GPIO_PIN15	= (1 shl 15);	//GPIO_7
VGI_GPIO_PIN_ALL= ($FFFF);    //ALL PINS

//函数声明
function VGI_ScanDevice(NeedInit:Byte):Integer; stdcall;
function VGI_OpenDevice(DevType,DevIndex,Reserved:Integer):Integer; stdcall;
function VGI_CloseDevice(DevType,DevIndex:Integer):Integer; stdcall;

function VGI_SetInput(DevType,DevIndex:Integer;Pins:Word):Integer; stdcall;
function VGI_SetOutput(DevType,DevIndex:Integer; Pins:Word):Integer; stdcall;
function VGI_SetOpenDrain(DevType,DevIndex:Integer; Pins:Word):Integer; stdcall;

function VGI_SetPins(DevType,DevIndex:Integer; Pins:Word):Integer; stdcall;
function VGI_ResetPins(DevType,DevIndex:Integer; Pins:Word):Integer; stdcall;

function VGI_ReadDatas(DevType,DevIndex:Integer;PinMask:Word;pData:PWord):Integer; stdcall;

implementation
function VGI_ScanDevice;external 'Ginkgo_Driver.dll' name 'VGI_ScanDevice';
function VGI_OpenDevice;external 'Ginkgo_Driver.dll' name 'VGI_OpenDevice';
function VGI_CloseDevice;external 'Ginkgo_Driver.dll' name 'VGI_CloseDevice';
function VGI_SetInput;external 'Ginkgo_Driver.dll' name 'VGI_SetInput';
function VGI_SetOutput;external 'Ginkgo_Driver.dll' name 'VGI_SetOutput';
function VGI_SetOpenDrain;external 'Ginkgo_Driver.dll' name 'VGI_SetOpenDrain';
function VGI_ResetPins;external 'Ginkgo_Driver.dll' name 'VGI_ResetPins';
function VGI_SetPins;external 'Ginkgo_Driver.dll' name 'VGI_SetPins';
function VGI_ReadDatas;external 'Ginkgo_Driver.dll' name 'VGI_ReadDatas';
end.
